### PR TITLE
WIP: add snapshot support to ScriptElements

### PR DESCRIPTION
**Pull request by adamraine**
To get script contents in snapshot mode, we need to fetch the contents over the protocol using `Debugger.getScriptSource` instead of looking at the network request. Unfortunatley, we can't get the `requestId` this way.

We can replace `ScriptElements` with the implementation in `ScriptElementsSnapshot` for timespan and navigation modes, but we'll need to add the `requestId` separately kinda like what `ImageRecords` does in https://github.com/GoogleChrome/lighthouse/pull/12663. Since `ScriptElements` and `ScriptElementsSnapshot` are so different, I'm inclined to make that a breaking change and keep two separate gatherers for now.

The main use case for `ScriptElementsSnapshot` would be a new audit `unminified-javascript-snapshot`, a snapshot version of `unminified-javascript`.

### CODE REVIEW

File: lighthouse-core/gather/gatherers/script-elements-snapshot.js [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+  /**
+   * @param {LH.Gatherer.FRTransitionalContext} context
+   * @return {Promise<LH.Artifacts['ScriptElements']>}
+   */
+  async _getArtifact(context) {
+    const session = context.driver.defaultSession;
+    const executionContext = context.driver.executionContext;
+
+    if (context.gatherMode === 'snapshot') {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/script-elements-snapshot.js:90]: it looks like you wrote this originally as a merged version, amirite? :)

**Review by patrickhulce — COMMENTED** [review]
Nice work!

> but we'll need to add the requestId separately kinda like what ImageRecords does in #12663.

FWIW, I think this is a positive thing to do regardless of a split. Aligns our artifacts as unique bits of data :)

> we need to fetch the contents over the protocol using Debugger.getScriptSource instead of looking at the network request

Does this work in our current gatherer in most cases? If so, I'm inclined to not maintain a fork. I don't know the answer to this but I might suspect that the network cache contents fetching we do is backed by the same data source as the `Debugger` version 🤷 

> The main use case for ScriptElementsSnapshot would be a new audit unminified-javascript-snapshot, a snapshot version of unminified-javascript.

Is this really it? If we're not able to convert any of the other JS opportunities as well, I'm also more hesitant to maintain a copy.

File: lighthouse-core/gather/gatherers/script-elements-snapshot.js (hunk 1 above)
Comment (adamraine) [reply lighthouse-core/gather/gatherers/script-elements-snapshot.js:90]: Guilty as charged

**Comment by adamraine** [comment]
> Does this work in our current gatherer in most cases?

It should. I didn't test it in navigation or timespan mode though.

> Is this really it?

That's what's possible with this change alone. I will need to reinvestigate some audits like `legacy-javascript` and `duplicated-javascript` which might work in snapshot mode if we can get `SourceMaps` to work in snapshot mode.

**Comment by adamraine** [comment]
> It should. I didn't test it in navigation or timespan mode though.

With a few changes, navigation and timespan are supported in the currently named `ScriptElementsSnapshot` gatherer.

File: lighthouse-core/gather/gatherers/script-elements-snapshot.js [hunk 2]
```diff
@@ -76,15 +82,14 @@ class ScriptElementsSnapshot extends FRGatherer {
   async stopSensitiveInstrumentation(context) {
     const session = context.driver.defaultSession;
     await session.sendCommand('Debugger.disable');
-    await session.off('Debugger.scriptParsed', this.onScriptParsed);
+    if (this.onScriptParsed) session.off('Debugger.scriptParsed', this.onScriptParsed);
```
Comment (connorjclark) [inline lighthouse-core/gather/gatherers/script-elements-snapshot.js:85]:
```suggestion
    if (this.onScriptParsed) await session.off('Debugger.scriptParsed', this.onScriptParsed);
```

Comment (adamraine) [reply lighthouse-core/gather/gatherers/script-elements-snapshot.js:85]: `session.off` and `session.on` return `void` no?

**Comment by patrickhulce** [comment]
> With a few changes, navigation and timespan are supported in the currently named ScriptElementsSnapshot gatherer.

Great! I'm a little confused though, seems like we don't need the separate gatherer in that case then? If so, what feedback are you blocked on from reviewer? :)

**Comment by adamraine** [comment]
> Great! I'm a little confused though, seems like we don't need the separate gatherer in that case then? If so, what feedback are you blocked on from reviewer? :)

The new gatherer cannot provide the `requestId`, so we would need to add it in a computed artifact like we did with `ImageElements`. Would this be a breaking change?

**Comment by patrickhulce** [comment]
> The new gatherer cannot provide the requestId, so we would need to add it in a computed artifact like we did with ImageElements. Would this be a breaking change?

Ah, yes it technically would. I couldn't find any usages with a quick GitHub, so we *could* consider returning `null` for all (semantic breaking change, but not a syntactical) 

Should we shelve this until we can make a few 9.0 breaking changes?

**Comment by patrickhulce** [comment]
> It's used in unminified-javascript and legacy-javascript.

I meant usages by other open source projects ;) we can definitely fix our own usages without breaking

> I'm ok with shelving, adding snapshot support here isn't mission critical.

Sounds good!

**Comment by connorjclark** [comment]
We will land this post-9.0, but probably keep existing `ScriptElements` intact until 10.0

**Pull request closed without merging**